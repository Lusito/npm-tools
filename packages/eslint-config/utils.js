// fixme: move react stuff to react?
const getReactRules = () => ({
    "react/jsx-filename-extension": [1, { extensions: [".tsx"] }],
    "react/destructuring-assignment": "off",
    "react/jsx-props-no-spreading": "off",
    "react/no-unescaped-entities": "off",
    "react/no-array-index-key": "off",
    "react/button-has-type": "off",
    "react-hooks/rules-of-hooks": "error",
    "react/sort-comp": "off",
    "react/require-default-props": "off",
    "react/function-component-definition": "off",
    "react/no-danger": "off",
    "react/react-in-jsx-scope": "off",
    "react/jsx-uses-react": "off",
});

module.exports = {
    getA11yOffRules() {
        // eslint-disable-next-line import/no-extraneous-dependencies
        return Object.keys(require("eslint-plugin-jsx-a11y").rules).reduce((acc, rule) => {
            acc[`jsx-a11y/${rule}`] = "off";
            return acc;
        }, {});
    },
    getConfig(react) {
        return {
            extends: [
                react ? "airbnb" : "airbnb-base",
                "plugin:import/errors",
                "plugin:import/warnings",
                "plugin:import/typescript",
                "prettier",
            ],
            plugins: ["import", ...(react ? ["react-hooks"] : [])],
            rules: {
                // Standard
                radix: "off",
                "global-require": "off",
                "no-param-reassign": "off",
                "no-restricted-syntax": "off",
                "default-case": "off",
                "no-console": "off",
                "consistent-return": "off",
                "no-plusplus": "off",
                "class-methods-use-this": "off",
                "no-continue": "off",
                camelcase: "off",
                // Import
                "import/no-unresolved": "off", // already done by tsc
                "import/order": [
                    "error",
                    {
                        groups: [
                            ["external", "internal", "builtin"],
                            ["parent", "sibling", "index"],
                        ],
                        "newlines-between": "always",
                    },
                ],
                "import/prefer-default-export": "off",
                "import/extensions": ["error", { ts: "never", tsx: "never" }],
                ...(react ? getReactRules() : {}),
            },
            overrides: [
                {
                    files: ["*.{ts,tsx}"],
                    extends: ["plugin:@typescript-eslint/recommended"],
                    parser: "@typescript-eslint/parser",
                    parserOptions: {
                        project: true,
                        tsconfigRootDir: ".",
                    },
                    plugins: ["@typescript-eslint"],
                    rules: {
                        // Rules handled by typescript
                        "no-shadow": "off",
                        "no-use-before-define": "off",
                        "no-unused-expressions": "off",
                        "@typescript-eslint/no-unused-expressions": [
                            "error",
                            { allowShortCircuit: true, allowTernary: true },
                        ],
                        "no-dupe-class-members": "off",
                        "@typescript-eslint/no-dupe-class-members": ["error"],
                        "@typescript-eslint/camelcase": ["off"],
                        // TypeScript
                        "@typescript-eslint/no-shadow": "error",
                        "@typescript-eslint/consistent-type-assertions": [
                            "warn",
                            { assertionStyle: "as", objectLiteralTypeAssertions: "allow" },
                        ],
                        "@typescript-eslint/no-explicit-any": "off",
                        "@typescript-eslint/explicit-function-return-type": "off",
                        "@typescript-eslint/type-annotation-spacing": "error",
                        "@typescript-eslint/array-type": ["error", { default: "array-simple" }],
                        "@typescript-eslint/default-param-last": "error",
                        "@typescript-eslint/explicit-module-boundary-types": "off",
                        "@typescript-eslint/naming-convention": [
                            "error",
                            {
                                selector: "typeParameter",
                                format: ["PascalCase"],
                                prefix: ["T"],
                                leadingUnderscore: "forbid",
                                trailingUnderscore: "forbid",
                            },
                            {
                                selector: "enumMember",
                                format: ["UPPER_CASE"],
                                leadingUnderscore: "forbid",
                                trailingUnderscore: "forbid",
                            },
                        ],
                        "@typescript-eslint/no-misused-promises": "error",
                        "@typescript-eslint/parameter-properties": "error",
                        // "@typescript-eslint/no-unnecessary-condition": "error",
                        "@typescript-eslint/prefer-as-const": "error",
                        "@typescript-eslint/no-unused-vars": ["error", { argsIgnorePattern: "^_" }],
                        "@typescript-eslint/prefer-for-of": "error",
                        "@typescript-eslint/prefer-includes": "error",
                        "@typescript-eslint/prefer-nullish-coalescing": "error",
                        "@typescript-eslint/prefer-optional-chain": "error",
                        "@typescript-eslint/restrict-plus-operands": "error",
                        "@typescript-eslint/restrict-template-expressions": "off",
                        "@typescript-eslint/no-use-before-define": [
                            "error",
                            { functions: false, typedefs: false, classes: false },
                        ],
                    },
                },
                {
                    files: ["*.js"],
                    parserOptions: {
                        ecmaVersion: "latest",
                    },
                    env: {
                        es6: true,
                    },
                },
                {
                    files: ["*.spec.{ts,tsx}"],
                    rules: {
                        "@typescript-eslint/no-non-null-assertion": "off",
                        "@typescript-eslint/no-empty-function": "off",
                        "@typescript-eslint/parameter-properties": "off",
                    },
                },
            ],
        };
    },
};
