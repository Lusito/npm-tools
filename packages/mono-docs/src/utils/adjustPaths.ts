export function applyAdjustPaths(path: string, adjustPaths: string[]) {
    for (const adjustPath of adjustPaths) {
        const parts = adjustPath.split("|");
        const newPath = path.replace(new RegExp(parts[0], "g"), parts[1]);
        if (path !== newPath) {
            return newPath;
        }
    }

    return path;
}
