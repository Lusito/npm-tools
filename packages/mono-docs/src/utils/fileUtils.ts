import { readdir, stat } from "fs/promises";
import { join } from "path";

export async function getAllFiles(dirPath: string, pattern: RegExp, arrayOfFiles: string[]) {
    const files = await readdir(dirPath);

    await Promise.all(
        files.map(async (file) => {
            const filePath = join(dirPath, file);
            const s = await stat(filePath);
            if (s.isDirectory()) {
                await getAllFiles(filePath, pattern, arrayOfFiles);
            } else if (pattern.test(file)) {
                arrayOfFiles.push(filePath);
            }
        })
    );

    return arrayOfFiles;
}
