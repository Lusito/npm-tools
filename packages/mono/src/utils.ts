import prompts from "prompts";
import type { PackageJson } from "type-fest";
import { readFile } from "fs/promises";
import { execSync } from "child_process";
import { resolve } from "path";

export type PublicPackageJson = PackageJson & {
    path: string;
    name: string;
    version: string;
    private: false;
};

export async function loadPackage(path: string) {
    const data = await readFile(path);

    return Object.assign(JSON.parse(data.toString()), { path }) as PackageJson;
}

export async function loadPackages() {
    const project = await loadPackage(resolve(process.cwd(), "./package.json"));

    const workspaceNames = Array.isArray(project.workspaces) ? project.workspaces : project.workspaces?.packages;

    if (!workspaceNames) throw new Error("No workspaces found");

    const workspaces = await Promise.all(
        workspaceNames.map((workspace) => loadPackage(resolve(process.cwd(), workspace, "package.json")))
    );

    return {
        project,
        workspaces,
    };
}

export async function prompt<T = string>(question: Omit<prompts.PromptObject<"value">, "name">): Promise<T> {
    const result = await prompts(
        {
            name: "value",
            ...question,
        },
        {
            onCancel() {
                console.error("Aborting");
                process.exit(-1);
            },
        }
    );

    return result.value;
}

export const run = (command: string) => execSync(command, { stdio: [0, 1, 2] });

export function die(message: string) {
    console.log("");
    console.error(`‚ùå ${message}`);
    process.exit(-1);
}
