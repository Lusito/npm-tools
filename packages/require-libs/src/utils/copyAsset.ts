import { copyFileSync, mkdirSync, readFileSync } from "fs";
import { createHash } from "crypto";
import { basename, extname, resolve } from "path";

export function createCopyAsset(assetDir: string, prefix: string) {
    const assetCache: Record<string, string | undefined> = {};
    mkdirSync(assetDir, { recursive: true });

    return function copyAsset(filename: string) {
        let destination = assetCache[filename];
        if (!destination) {
            const fileBuffer = readFileSync(filename);
            const hashSum = createHash("sha256");
            hashSum.update(fileBuffer);
            const hash = hashSum.digest("hex").substring(0, 8);
            const ext = extname(filename);

            const newFilename = `${basename(filename, ext)}-${hash}${ext}`;

            destination = prefix + newFilename;
            assetCache[filename] = destination;

            copyFileSync(filename, resolve(assetDir, newFilename));
        }
        return destination;
    };
}
