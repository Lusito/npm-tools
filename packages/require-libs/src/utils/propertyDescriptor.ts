export function propertyDescriptor(value: unknown): PropertyDescriptor {
    return {
        enumerable: false,
        configurable: false,
        writable: false,
        value,
    };
}
