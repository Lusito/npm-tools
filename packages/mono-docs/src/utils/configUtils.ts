import { dirname, join, relative, resolve } from "path";
import { existsSync } from "fs";
import fs from "fs/promises";
import yaml from "js-yaml";

import { getAllFiles } from "./fileUtils";
import { CombinedDocsConfig, DocsConfig } from "./types";

type LoadedConfig = {
    dir: string;
    file: string;
    depth: number;
    data: DocsConfig;
};

async function loadConfig(rootPath: string, file: string): Promise<LoadedConfig> {
    const content = await fs.readFile(file, "utf-8");
    const data = yaml.load(content) as DocsConfig;

    const relativePath = relative(rootPath, file);
    const dir = dirname(relativePath);

    return {
        dir,
        file,
        depth: relativePath.split("/").length,
        data,
    };
}

const defaultDocsConfig: CombinedDocsConfig = {
    siteName: "Unnamed Documentation",
    description: "",
    footer: [],
    keywords: [],
    links: [],
    projects: [],
    sidebar: [],
    title: "",
    adjustPaths: [],
    docs: "docs",
};

function adjustConfig(rootPath: string, loadedConfig: LoadedConfig) {
    const { dir, data } = loadedConfig;
    const config = data as CombinedDocsConfig;
    const { sidebar, projects, docs } = config;
    const basePath = resolve(rootPath, dir);
    const docsPath = resolve(basePath, docs);
    data.docs = docsPath;

    config.sidebar = sidebar.map((entry) => join(docsPath, `${entry}.md`));

    if (config.sidebar.length) {
        const readme = join(basePath, "README.md");
        if (existsSync(readme)) {
            config.sidebar.unshift(readme);
        }
    }
    config.projects = projects.map((project) => join(rootPath, `${project}/README.md`));
}

export type ConfigGetter = (dir: string) => CombinedDocsConfig;
export type LoadedConfigs = {
    rootConfig: CombinedDocsConfig;
    getConfig: ConfigGetter;
};

export async function loadConfigs(rootPath: string): Promise<LoadedConfigs> {
    const configFiles = await getAllFiles(rootPath, /^mono-docs\.yml$/, []);
    const configs = await Promise.all(configFiles.map((file) => loadConfig(rootPath, file)));

    configs.sort((a, b) => a.depth - b.depth);

    // Perform inheritance from parent configs
    const rootConfig = configs[0];
    rootConfig.data = { ...defaultDocsConfig, ...rootConfig.data };

    for (let i = 1; i < configs.length; i++) {
        const config = configs[i];
        const possibleParentConfigs = configs.filter((c) => config.dir.startsWith(c.dir));
        if (possibleParentConfigs.length > 1) {
            // configs are sorted by depth, the last one is the config itself and the second-last is the parent
            const parentConfig = possibleParentConfigs[possibleParentConfigs.length - 2];
            config.data = { ...parentConfig.data, ...config.data };
        } else {
            config.data = { ...rootConfig.data, ...config.data };
        }
    }

    for (const config of configs) {
        adjustConfig(rootPath, config);
    }

    return {
        rootConfig: rootConfig.data as CombinedDocsConfig,
        getConfig(dir: string) {
            const possibleConfigs = configs.filter((c) => dir.startsWith(c.dir));
            // configs are sorted by depth, the last one is the best matching config
            const config = possibleConfigs[possibleConfigs.length - 1];
            if (config) {
                return config.data as CombinedDocsConfig;
            }

            return rootConfig.data as CombinedDocsConfig;
        },
    };
}
