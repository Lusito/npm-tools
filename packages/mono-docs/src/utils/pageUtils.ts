import { basename, dirname, relative, resolve } from "path";
import slugify from "slug";
import { MarkdownModule } from "@lusito/require-libs";

import { getAllFiles } from "./fileUtils";
import { CombinedDocsConfig, PageInfo, PageMeta } from "./types";
import { applyAdjustPaths } from "./adjustPaths";
import { ConfigGetter } from "./configUtils";

function slugFromFilename(filename: string) {
    const slug = slugify(basename(filename, ".md"));

    return slug === "readme" ? "index" : slug;
}

async function loadPage(rootPath: string, file: string, getConfig: ConfigGetter): Promise<PageInfo> {
    const data: MarkdownModule = await import(relative(__dirname, file));

    const relativePath = relative(rootPath, file);
    const dir = dirname(relativePath);
    const pathPrefix = dir === "." ? "" : `/${dir.replaceAll("\\", "/")}`;
    const slug = slugFromFilename(relativePath);

    const key = slug === "index" ? pathPrefix : `${pathPrefix}/${slug}`;

    const docsConfig = getConfig(dir);

    return {
        dir,
        file,
        depth: key.split("/").length,
        key,
        meta: data.meta as PageMeta,
        docsConfig,
        path: applyAdjustPaths(`${pathPrefix}/${slug}.html`, docsConfig.adjustPaths),
        body: data.html,
    };
}

async function getDocsPages(rootPath: string, docsPath: string, getConfig: ConfigGetter) {
    const files = await getAllFiles(docsPath, /\.md$/, []);
    const pages = await Promise.all(files.map((file) => loadPage(rootPath, file, getConfig)));
    pages.sort((a, b) => a.depth - b.depth);

    return pages;
}

async function getProjectPages(rootPath: string, project: string, getConfig: ConfigGetter) {
    const projectPath = dirname(project);
    const indexPage = await loadPage(rootPath, project, getConfig);
    indexPage.projectIndex = indexPage;

    const projectConfig = getConfig(relative(rootPath, projectPath));

    const pages = await getDocsPages(rootPath, projectConfig.docs, getConfig);

    for (const page of pages) {
        page.projectIndex = indexPage;
    }

    return [indexPage, ...pages];
}

export async function getPages(
    rootPath: string,
    rootConfig: CombinedDocsConfig,
    getConfig: ConfigGetter
): Promise<PageInfo[]> {
    const rootPage = await loadPage(rootPath, resolve(rootPath, "README.md"), getConfig);

    // Monorepo setup
    if (rootConfig.projects.length) {
        const subPages = await Promise.all(
            rootConfig.projects.map((project) => getProjectPages(rootPath, project, getConfig))
        );

        return [rootPage, ...subPages.flat()];
    }

    // Non-monorepo setup
    const pages = await getDocsPages(rootPath, rootConfig.docs, getConfig);
    rootPage.projectIndex = rootPage;

    return [rootPage, ...pages];
}

export const getPageTitle = (currentPage: PageInfo) => {
    const { docsConfig, projectIndex, meta } = currentPage;
    const title = projectIndex === currentPage ? docsConfig.title || meta.title : meta.title || docsConfig.title;
    return title || docsConfig.siteName;
};
