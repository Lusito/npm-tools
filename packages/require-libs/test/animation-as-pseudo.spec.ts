import { css, describeBlock } from "./utils";

const source = css`
    @keyframes hover {
        0% {
            opacity: 0;
        }

        100% {
            opacity: 0.5;
        }
    }

    @media (max-width: 480px) {
        .animation:hover {
            background: green;
        }
    }
`;

describe.each([
    ["unscoped", false, source],
    ["scoped", true, source],
])("animation-as-pseudo (%s)", describeBlock);
