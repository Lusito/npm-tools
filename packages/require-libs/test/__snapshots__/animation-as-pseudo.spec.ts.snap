// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`animation-as-pseudo (scoped) produces the correct css 1`] = `
"@keyframes hover_Zll8S {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 0.5;
  }
}
@media (max-width: 480px) {
  .animation_Zll8S:hover {
    background: green;
  }
}"
`;

exports[`animation-as-pseudo (scoped) produces the correct file mappings 1`] = `[]`;

exports[`animation-as-pseudo (scoped) produces the correct tokens 1`] = `
{
  "animation": "animation_Zll8S",
  "hover": "hover_Zll8S",
}
`;

exports[`animation-as-pseudo (unscoped) produces the correct css 1`] = `
"@keyframes hover {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 0.5;
  }
}
@media (max-width: 480px) {
  .animation:hover {
    background: green;
  }
}"
`;

exports[`animation-as-pseudo (unscoped) produces the correct file mappings 1`] = `[]`;

exports[`animation-as-pseudo (unscoped) produces the correct tokens 1`] = `{}`;
